// @kintone/rest-api-client/lib/platform/node.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVersion = exports.buildBaseUrl = exports.buildFormDataValue = exports.buildHeaders = exports.buildPlatformDependentConfig = exports.getDefaultAuth = exports.getRequestToken = exports.readFileFromPath = void 0;
var fs_1 = __importDefault(require("fs"));
var util_1 = require("util");
var path_1 = require("path");
var UnsupportedPlatformError_1 = require("./UnsupportedPlatformError");
var https_1 = __importDefault(require("https"));
var os_1 = __importDefault(require("os"));
var packageJson = require("../../package.json");
var readFile = (0, util_1.promisify)(fs_1.default.readFile);
var readFileFromPath = function (filePath) { return __awaiter(void 0, void 0, void 0, function () {
    var data, name;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, readFile(filePath)];
            case 1:
                data = _a.sent();
                name = (0, path_1.basename)(filePath);
                return [2 /*return*/, { data: data, name: name }];
        }
    });
}); };
exports.readFileFromPath = readFileFromPath;
var getRequestToken = function () {
    throw new UnsupportedPlatformError_1.UnsupportedPlatformError("Node.js");
};
exports.getRequestToken = getRequestToken;
var getDefaultAuth = function () {
    throw new UnsupportedPlatformError_1.UnsupportedPlatformError("Node.js");
};
exports.getDefaultAuth = getDefaultAuth;
var buildPlatformDependentConfig = function (_a) {
    var httpsAgent = _a.httpsAgent, clientCertAuth = _a.clientCertAuth;
    if (httpsAgent !== undefined) {
        return { httpsAgent: httpsAgent };
    }
    // use default HTTPS agent
    if (clientCertAuth !== undefined) {
        var pfx = "pfx" in clientCertAuth
            ? clientCertAuth.pfx
            : fs_1.default.readFileSync(clientCertAuth.pfxFilePath);
        var defaultHttpsAgent = new https_1.default.Agent({
            pfx: pfx,
            passphrase: clientCertAuth.password,
        });
        return { httpsAgent: defaultHttpsAgent };
    }
    return {};
};
exports.buildPlatformDependentConfig = buildPlatformDependentConfig;
var buildHeaders = function (params) {
    var userAgent = params.userAgent;
    return {
        "User-Agent": "Node.js/".concat(process.version, "(").concat(os_1.default.type(), ") ").concat(packageJson.name, "@").concat(packageJson.version).concat(userAgent ? " ".concat(userAgent) : ""),
    };
};
exports.buildHeaders = buildHeaders;
var buildFormDataValue = function (data) {
    return data;
};
exports.buildFormDataValue = buildFormDataValue;
var buildBaseUrl = function (baseUrl) {
    if (typeof baseUrl === "undefined") {
        throw new Error("in Node.js environment, baseUrl is required");
    }
    return baseUrl;
};
exports.buildBaseUrl = buildBaseUrl;
var getVersion = function () {
    return packageJson.version;
};
exports.getVersion = getVersion;
